pkg/sysched/sysched.go

```diff
@@ -0,0 +1,505 @@
+		// NOTE: should we consider the rest categories, e.g., notify, trace?
+		// SCMP_ACT_TRACE --> ActTrace, seccomp.ActNotify
+		if element.Action == seccomp.ActAllow || element.Action == seccomp.ActLog {
+			syscalls = unionList(syscalls, element.Names)
+		}
+	}
+
+	return syscalls, nil
+}
+
+// obtains the system call list for a pod from the pod's seccomp profile
+// SPO is used to generate and input the seccomp profile to a pod
+// If a pod does not have a SPO seccomp profile, then an unconfined
+// system call set is return for the pod
+func (sc *SySched) getSyscalls(pod *v1.Pod) []string {
+	var r []string
+
+	// read the seccomp profile from the security context of a pod
+	podSC := pod.Spec.SecurityContext
+	if podSC != nil && podSC.SeccompProfile != nil && podSC.SeccompProfile.Type == "Localhost" {
+		if podSC.SeccompProfile.LocalhostProfile != nil {
+			profilePath := *podSC.SeccompProfile.LocalhostProfile
+			ns, crdName := getCRDandNamespace(profilePath)
+
+			if len(ns) > 0 && len(crdName) > 0 {
+				syscalls, err := sc.readSPOProfileCRD(crdName, ns)
+				if err != nil {
+					klog.ErrorS(err, "Failed to read syscall CRD by parsing pod security context")
+				}
+
+				if len(syscalls) > 0 {
+					r = unionList(r, syscalls)
+				}
+			}
+		}
+	}
+
+	// read the seccomp profile from container security context and merge them
+	for _, container := range pod.Spec.Containers {
+		conSC := container.SecurityContext
+		if conSC != nil && conSC.SeccompProfile != nil && conSC.SeccompProfile.Type == "Localhost" {
+			if conSC.SeccompProfile.LocalhostProfile != nil {
+				profilePath := *conSC.SeccompProfile.LocalhostProfile
+				ns, crdName := getCRDandNamespace(profilePath)
+
+				if len(ns) > 0 && len(crdName) > 0 {
+					syscalls, err := sc.readSPOProfileCRD(crdName, ns)
+					if err != nil {
+						klog.ErrorS(err, "Failed to read syscall CRD by parsing container security context")
+					}
+
+					if len(syscalls) > 0 {
+						r = unionList(r, syscalls)
+					}
+				}
+			}
+		}
+	}
+
+	// SPO seccomp profiles are sometimes automatically annotated to a pod
+	if pod.ObjectMeta.Annotations != nil {
+		// there could be multiple SPO seccomp profile annotations for a pod
+		// merge all profiles to obtain the syscal set for a pod
+		for k, v := range pod.ObjectMeta.Annotations {
+			// looks for annotation related to the seccomp
+			if strings.Contains(k, SPO_ANNOTATION) {
+				ns, crdName := getCRDandNamespace(v)
+
+				if len(ns) > 0 && len(crdName) > 0 {
+					syscalls, err := sc.readSPOProfileCRD(crdName, ns)
+
+					if err != nil {
+						klog.ErrorS(err, "Failed to read syscall CRD by parsing pod annotation")
+						continue
+					}
+
+					if len(syscalls) > 0 {
+						r = unionList(r, syscalls)
+					}
+				}
+			}
+		}
+	}
+
+	// if a pod does not have a seccomp profile specified, return the set of all syscalls
+	if len(r) == 0 {
+		syscalls, err := sc.readSPOProfileCRD(sc.DefaultProfileName, sc.DefaultProfileNamespace)
```

Comment:
Is this line covered by unit test? If no, could you craft a test to ensure len(r) == 0 so test reaches here?